gennet-module/gennet.py

```diff
@@ -256,33 +287,79 @@ def dict_to_arrays(dic):
     return keys, vals
 
 
+# Check for range failures in a list
+def range_fails(lst, min=0, max=100):
+    return any(x < min or x > max for x in lst)
+
+
+# Check for sum failures in a list
+def sum_fails(lst, sum_expected=100):
+    return not sum(lst) == sum_expected
+
+
+# Construct the nodeType from the trait
+def traits_to_nodeType(s):
+    return nodeType(s.split(':')[0])
+
+# Validate the traits distribution (stick to percentages: num nodes may vary post generation)
+def validate_traits_distribution(traits_distribution, num_nodes):
+    traits, traits_freq = dict_to_arrays(traits_distribution)
+    if range_fails(traits_freq, max=100) :
+        raise ValueError(f"{traits_distribution} : invalid percentage (>{100} or <0)")
+    if sum_fails(traits_freq, sum_expected=100) :
+        raise ValueError(
+                f"{traits_distribution} : percentages do not sum to {100}")
+    for s in traits :
+        traits_list = s.split(":")
+        for  t in traits_list :
+            print(f"{TRAITS_DIR}/{t}.toml")
+            if t not in Trait and not os.path.exists(f"{TRAITS_DIR}/{t}.toml") :
+                raise ValueError(
+                    f"{traits_distribution} : unknown trait {t} in {s}")
+
+
+# Extract the traits distribution
+def extract_traits_distribution(traits_distribution):
+    return dict_to_arrays(traits_distribution)
+
+
 # Generate a list of nodeType enums that respects the node type distribution
-def generate_node_types(node_type_distribution, G):
+def generate_traits_distribution(node_type_distribution, G):
     num_nodes = G.number_of_nodes()
-    nodes, node_probs = dict_to_arrays(node_type_distribution)
-    node_types_str = random.choices(nodes, weights=node_probs, k=num_nodes)
-    node_types_enum = [nodeType(s) for s in node_types_str]
-    return node_types_enum
+    nodes, node_percentage = dict_to_arrays(node_type_distribution)
+    traits_distribution = []
+    for i, n in enumerate(nodes) :
+       traits_distribution +=  [nodes[i]] * math.ceil(node_percentage[i] * num_nodes/100)
+    random.shuffle(traits_distribution)
+    return traits_distribution
+
 
-# Inverts a dictionary of lists
-def invert_dict_of_list(d):
+# Inverts a dictionary of lists (of lists/tuples) 
+def invert_dict_of_list(d, idx=0):
     inv = {}
-    for key, val in d.items():
-        if val not in inv:
-            inv[val] = [key]
-        else:
-            inv[val].append(key)
+    for key, val in d.items() :
+        if idx == 0 :
+            if val not in inv :
+                inv[val] = [key]
+            else :
+                inv[val].append(key)
+        else :
+            if val[1] not in inv :
+                inv[val[1]] = [key]
+            else :
+                inv[val[1]].append(key)
```

Comment:
This is better written as:

```python
from collections import defaultdict
def inver_dict_of_list(d, idx: bool):
    inv = defaultdict(list)
    idx = 0 if idx else 1
    for key, val in d.items():
        inv[val[idx]].appen(val)
    return inv
```

Reply:
Also, it can be done as:
```python
reversed_dict = defaultdict(list)
for key, value in d.items():
    reversed_dict[value].append(key)
```    

But I think this is a but unnecessary if `generate_subnets` could create this in the correct form from the beginning

Reply:
that indeed reads much better, thanks! updated the branch.

> This is better written as:
> 
> ```python
> from collections import defaultdict
> def inver_dict_of_list(d, idx: bool):
>     inv = defaultdict(list)
>     idx = 0 if idx else 1
>     for key, val in d.items():
>         inv[val[idx]].append(val)
>     return inv
> ```